rules/readfile.go

```diff
@@ -0,0 +1,50 @@
+// (c) Copyright 2016 Hewlett Packard Enterprise Development LP
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rules
+
+import (
+	"go/ast"
+	"go/types"
+
+	"github.com/GoASTScanner/gas"
+)
+
+type readfile struct {
+	gas.CallList
+}
+
+// Match inspects AST nodes to determine if the match the methods `os.Open` or `ioutil.ReadFile`
+func (r *readfile) Match(n ast.Node, c *gas.Context) (*gas.Issue, error) {
+	if node := r.ContainsCallExpr(n, c); node != nil {
+		for _, arg := range node.Args {
+			if ident, ok := arg.(*ast.Ident); ok {
+				obj := c.Info.ObjectOf(ident)
+				if _, ok := obj.(*types.Var); ok && !gas.TryResolve(ident, c) {
+					return gas.NewIssue(c, n, "File inclusion launched with variable", gas.Medium, gas.High), nil
+				}
+			}
+		}
+		return gas.NewIssue(c, n, "File inclusion launching should be audited", gas.Low, gas.High), nil
```

Comment:
Isn't a bit to much this warning. This will trigger an issues for all files regardless if they are provided as a taint input or not. This will produce potentially to many false positive warnings.

Reply:
I followed the same pattern for the `subproc` rule.  What should I do in this case?

Does GoAST have a lib that can do some data-flow analysis that pinpoints where the variable was assigned and possibly tainted? I played around with the `Context` struct but wasn't able to get the results I wanted.

Reply:
> I followed the same pattern for the subproc rule. What should I do in this case?

I would return nil instead.

> Does GoAST have a lib that can do some data-flow analysis that pinpoints where the variable was assigned and possibly tainted? I played around with the Context struct but wasn't able to get the results I wanted.

I found this package which does some data flow analysis https://github.com/godoctor/godoctor/tree/master/analysis/dataflow. It would be great if you could try it out and come up with a more generic solution for taint input vulnerability.